Corona-Warn-App/src/main/java/de/rki/coronawarnapp/vaccination/core/repository/ValueSetsRepository.kt

```diff
@@ -19,55 +26,78 @@ import javax.inject.Inject
-    @AppScope private val scope: CoroutineScope
+    @AppScope private val scope: CoroutineScope,
+    dispatcherProvider: DispatcherProvider
 ) {
 
-    val latestValueSet: Flow<VaccinationValueSet> = valueSetsStorage.valueSet.flow.distinctUntilChanged()
+    private fun Flow<VaccinationValueSet>.distinctUntilChangedByHash() = distinctUntilChangedBy { it.hashCode() }
 
-    fun triggerUpdateValueSet(languageCode: Locale) = scope.launch {
+    private val internalData: HotDataFlow<VaccinationValueSet> = HotDataFlow(
+        loggingTag = TAG,
+        scope = scope,
+        coroutineContext = dispatcherProvider.IO,
+        sharingBehavior = SharingStarted.Lazily,
+        startValueProvider = {
+            (valueSetsStorage.vaccinationValueSet ?: createEmptyValueSet())
+                .also { Timber.v("Loaded initial value set %s", it) }
+        }
+    )
+
+    init {
+        internalData.data
+            .distinctUntilChangedByHash()
+            .onStart { Timber.d("Observing value set") }
+            .onEach { valueSetsStorage.vaccinationValueSet = it }
+            .catch { Timber.e(it, "Storing new value set failed.") }
+            .launchIn(scope + dispatcherProvider.IO)
+    }
+
+    val latestValueSet: Flow<VaccinationValueSet> = internalData.data.distinctUntilChangedByHash()
+
+    fun triggerUpdateValueSet(languageCode: Locale) {
         Timber.d("triggerUpdateValueSet(languageCode=%s)", languageCode)
+
+        internalData.updateAsync(
+            onUpdate = { updateValueSet(languageCode = languageCode) },
+            onError = { Timber.e(it, "Updating value set failed") }
+        )
+    }
+
+    private suspend fun updateValueSet(languageCode: Locale): VaccinationValueSet {
+        Timber.v("updateValueSet(languageCode=%s)", languageCode)
         var valueSet = vaccinationServer.getVaccinationValueSets(languageCode = languageCode)
+            ?: valueSetsStorage.vaccinationValueSet
 
-        if (valueSet == null && valueSetsStorage.valueSet.value.isEmpty()) {
+        if (valueSet.isEmpty()) {
             Timber.d(
-                "Got no value set from server for %s and local value set is empty... " +
-                    "Try fallback to value set for en",
+                "Got no value set from server for %s and local value set is empty... Try fallback to value set for en",
                 languageCode.language
             )
             valueSet = vaccinationServer.getVaccinationValueSets(languageCode = Locale.ENGLISH)
         }
 
-        valueSet
-            .also { Timber.d("Value set is %s", it) }
-            ?.let { newValueSet -> valueSetsStorage.valueSet.update { newValueSet.toStoredVaccinationValueSet() } }
+        return when (valueSet != null) {
+            true -> valueSet
+            else -> createEmptyValueSet() // Fallback to empty value set
+        }.also { Timber.v("New value set %s", it) }
     }
 
-    fun clear() {
+    private fun createEmptyValueSet(): VaccinationValueSet =
```

Comment:
Could be a static value (i.e. companion object)?

Reply:
all right. will adjust